SFDigitalServices/sfgov
# Sg 1054 - Form.io block and Feedback form

## Summary of changes

- Implements a new Feedback form that appears at the bottom of every page.
- Adds a custom block entity named "Form" which utilizes the Form.io paragraph (and can support other types in the future as needed).
- Adds a custom module for Form.io to store library definitions and customizations, and documentation for how to use within Drupal.
- Tweaks the existing template implementation to support multiple forms per page, and Twig blocks (so we have less code repetition where customizations are required).
- Provides a fixed content block instance, configured to print in the `content_bottom` region. This block loads on every page.  
- Removes the ability to define a version per Form.io form (see additional notes for more details).
  ![Screenshot of Form.io Paragraph form](https://user-images.githubusercontent.com/60979/83542829-8382f180-a4c9-11ea-9c70-b5493730cdcf.png)
- Adds a configuration page that allows changing the source of the Form.io and SDFS scripts.
    ![Screenshot of admin config](https://user-images.githubusercontent.com/60979/83542945-aad9be80-a4c9-11ea-912c-8672ecb7c90f.png)


<details>
<summary>Additional Notes</summary>

1. I had to remove the fields and code that allows defining versioning per-form, because the second two forms are on a page, which with this PR, will be the case for every [form page](https://sfgov.gov/workers-families-first-preapproval-form), you risk loading multiple versions of both Form.io and [formio-sfds](https://github.com/SFDigitalServices/formio-sfds). I have replaced this with a global configuration. See [this comment](https://github.com/SFDigitalServices/sfgov/pull/534#discussion_r434003884) for details.
2. Despite not needing [formio-sfds](https://github.com/SFDigitalServices/formio-sfds) for the Feedback form, I had to load it anyway, because like above, if it's loaded on the page, for another form, it will be initialized, which changes a lot about the form's HTML structure and necessary CSS. Also, I realized that the SFDS theme is required for i18n functionality to work.
3. I was unable to get the forms to work using `async`.  If anyone has ideas, lmk!
4. I attempted to use a locally-hosted version of the form config, but could not figure out how to  "hook" it back up to Form.io.  <del>I left the implementation code in here, which is harmless, in case anyone wishes to try it out.  Given an endpoint of `https://sfds.form.io/workers-families-first` it will look for `web/modules/custom/sfgov_formio/data/workers-families-first.json`, and if found, it will pass the `config` as JSON directly to `Formio.createForm`</del>.

</details>

## Testing Notes

- Visit https://pr-534-sfgov.pantheonsite.io/. The Feedback form should be at the bottom of the each page.
- Visit https://pr-534-sfgov.pantheonsite.io/workers-families-first-preapproval-form, which is an example of a page that has two instances on a page.
- Visit https://pr-534-sfgov.pantheonsite.io/node/973, which is the building permit form (currently under development)

## Deployment Notes

Configure the fixed content block's content:

1. Load the site (e.g. homepage). This creates the fixed block content instance.
1. Log in, and go to `/block/2`.
2. Click "Add Form.io"
3. In "Data Source" field, add: `https://sfds-test.form.io/feedback `
4. In "Form.io Render Options" field, add: 

    ```json
    {
      "formioSFDSOptOut": true,
      "i18n": "https://i18n-microservice-js.herokuapp.com/google/1ialzMAruk1wk2sIAR-WHzlCsJ73yy7LIxruD3cvTYFE"
    }
    ```
5. Save.

You should now see the form at the bottom of each page.

<img width="1443" alt="feedback-form" src="https://user-images.githubusercontent.com/60979/81977824-9efc8a00-95f8-11ea-8378-f8a999e9af75.png">

## Todo

- [x] Get help testing existing forms, in the hopes of removing the version fields entirely (they are not functional with this PR anyway).
- [x] Get proper access to Airtable and Form.io to move the form and data outside of my personal account ;)
